File: apps/spring/src/main/java/com/benchmark/spring/controller/BenchmarkController.java
package com.benchmark.spring.controller;

import com.benchmark.spring.entity.User;
import com.benchmark.spring.service.BenchmarkService;
import java.util.List;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api")
public class BenchmarkController {

  private final BenchmarkService benchmarkService;

  public BenchmarkController(BenchmarkService benchmarkService) {
    this.benchmarkService = benchmarkService;
  }

  @GetMapping("/health")
  public Map<String, String> health() {
    return benchmarkService.health();
  }

  @GetMapping("/compute")
  public Map<String, Long> compute(@RequestParam(defaultValue = "40") int n) {
    return Map.of("result", benchmarkService.compute(n));
  }

  @GetMapping("/users")
  public List<User> getUsers() {
    return benchmarkService.getUsersOrm();
  }

  @GetMapping("/users/raw")
  public List<Map<String, Object>> getUsersRaw() {
    return benchmarkService.getUsersRaw();
  }

  @PostMapping("/users")
  @ResponseStatus(HttpStatus.CREATED)
  public User createUser(@RequestBody User user) {
    return benchmarkService.createUser(user);
  }
}

File: apps/nestjs/src/benchmark/benchmark.controller.ts
import { Body, Controller, Get, Post, Query } from '@nestjs/common';
import type { User } from '../users/user.entity';
import { BenchmarkService } from './benchmark.service';

@Controller()
export class BenchmarkController {
  constructor(private readonly benchmarkService: BenchmarkService) {}

  @Get('health')
  health() {
    return this.benchmarkService.health();
  }

  @Get('compute')
  compute(@Query('n') n = '40') {
    return { result: this.benchmarkService.compute(Number(n)) };
  }

  @Get('users')
  async getUsers(): Promise<Array<{ id: number; name: string; email: string; created_at: Date }>> {
    const users = await this.benchmarkService.getUsersOrm();
    return users.map((user) => ({
      id: user.id,
      name: user.name,
      email: user.email,
      created_at: user.createdAt,
    }));
  }

  @Get('users/raw')
  getUsersRaw(): Promise<any[]> {
    return this.benchmarkService.getUsersRaw();
  }

  @Post('users')
  async createUser(
    @Body() body: { name: string; email: string },
  ): Promise<{ id: number; name: string; email: string; created_at: Date }> {
    const user = await this.benchmarkService.createUser(body);
    return {
      id: user.id,
      name: user.name,
      email: user.email,
      created_at: user.createdAt,
    };
  }
}
